### ISSUE TITLE
Roadmap To 1.0 And Beyond

### ISSUE THREAD

**Issue by thescientist13** [ProjectEvergreen/greenwood#418]
This issue is intended to outline a the short and long term planning of the project for review and consensus gathering.  It is broken down into the following milestones as part of the discovery done as part of the [NoNo POC](https://github.com/thescientist13/nono-poc) and the work to port that to Greenwood here in #417.

This will also serve as a way to update current GitHub Issues, Projects, Milestones

## Questions
- [`<meta>` component / package](https://github.com/ProjectEvergreen/greenwood/issues/304) still needed? - yes!  have some **react-helmet** like features
- [RFC Declarative App templates](https://github.com/ProjectEvergreen/greenwood/issues/299) still needed? - no, not really I don't think

## Feature Parity (relatively)  by 12/31/2020
This is work needed to accomplish the bulk of the work necessary to get #355 merged without much disruption to existing users.  Merged into an ongoing `release/1.0.0` release branch and tracked in the Architecture Digest Project / MVP Milestone.

Will want to validate these using an `rc` or `next` branch release before merging into `master`
* Greenwood Getting Started - https://github.com/ProjectEvergreen/greenwood-getting-started/pull/22
* www.thegreenhouse.io - https://github.com/thegreenhouseio/www.thegreenhouse.io/pull/182
* projectevergreen.github.io - https://github.com/ProjectEvergreen/projectevergreen.github.io/pull/36

Main thing here should be ensure that performance improves significantly and a number of existing issues in the backlog can be resolved.  Each of the below numbered items would be a PR and tracked in the release branch PR.

1. ~~PR and associated TODOs~~ from #417 
1. ~~Refactor _serve.js_ into plugin based architecture~~ bring done as part of #185 
1. ~~SPA support~~ - delegated to #354 
    - (how to) handle strict / ssg `optimization` in _greenwood.config.js_
    - Decouple spa specific code from the bundle
    - [Smaller / lighter router](https://router.matsu.fi/getting-started)?
    - [rollup dynamic input vars](https://github.com/rollup/plugins/tree/master/packages/dynamic-import-vars)
    - puppeteer optional?  e.g. SPA no SSG, SSG, vs SPA vs Hybrid (strict?)
1. ~~[GraphQL as a package](https://github.com/ProjectEvergreen/greenwood/issues/278)~~, or [something simpler](https://jaredpalmer.com/gatsby-vs-nextjs)
    - update all usages to use `@data/client` instead of `fetch`
    - [data hydration still an issue](https://github.com/ProjectEvergreen/greenwood/issues/349)
    - [reduce duplicate graphql calls](https://github.com/ProjectEvergreen/greenwood/issues/272)?
1. ~~Handle loading prismjs~~ - delegated to #427
1. ~~How to handle user context options / default CLI _templates/_~~ - tracked in #428 
1. ~~Refactor _serve.js_ into middleware based design~~ - #519 
1.  ~~create [JSON and CSS plugins using `import.meta.url`](https://twitter.com/dassurma/status/1358819452708659204)~~ - see #185
    - created issue for JSON plugin - https://github.com/ProjectEvergreen/greenwood/issues/516
    - created issue for image plugin - https://github.com/ProjectEvergreen/greenwood/issues/517
1. ~~Smarter handling of `<title></title>` tag, takes a function?~~ - deferred to #304 
1. ~~!Restore Polyfill and Google Plugins~~ - will be deferred to #185 
1. ~~Babel / PostCSS / Browserslist support~~ - created #426
    - restore nested styles
    - enable for dev mode
1. ~~Enable / get all specs passing~~ - tracked in #428
1. ~~Rollup refactoring~~ - tracked #428
    - bundle everything from _www/_?
    - how to handle empty chunks warning from rollup?
    - avoid _style.js_ file being generated by rollup CSS plugin
1. ~~[figure out why I can't use pages/blog/index.html](https://github.com/ProjectEvergreen/greenwood/issues/120)~~ fixed by #417 
1. ~~confirm [`body[unresolved]` is still an issue](https://github.com/thescientist13/nono-poc/pull/5#issuecomment-690810774)~~ - see #433 
1. ~~confirm `import-map` support shim needed (for local dev)~~ #485
1. ~~[handling paths in templates, e.g. ./, ../](https://github.com/ProjectEvergreen/greenwood/issues/273), normalize everything to `/`?, also associated with [route misses with using /](https://github.com/ProjectEvergreen/greenwood/issues/331)~~ - tracked in #435
1. ~~Revisit all documentation~~ tracked in #430
1. ~~Test more dependencies to ensure proper handling of CJS vs ESM based~~ - tracked in #429
1. ~~Publish to next branch and test in critical consumer projects~~ - #485

## Roadmap to 1.0 by 12/31/2021
Things we should try and complete by end of year to achieve a 1.0.  The aim is to complete low effort features that introduce / exercise new APIs, rounds out certain key features needed as part of the Feature Parity work.

1. [x] Noticed that the CSS **puppeteer** (and whitespace from us?) puts into an _index.html_ page is almost `10KB`!!!  This _may_ be causing style collisions colliding with styles 👆 ?  Either way, we should control the HTML output.  If want to support inlining CSS great, but then we also shouldn't ship it over the network either.  One _or_ the other.  Will make an issue to track this specific issue of junk in the trunk, and another to track optimizing for inlining CSS (may not work "work" for CSS-in-JS or else you get the FOUC thought...). - see #453 for more context
    - will leave this the way it is for now
1. [x] Right now we have to process everything twice to support the **user**'s PostCSS / Babel (once in serve and again in bundle / rollup since we read from file, not the server) in Rollup .  It would be nice if Rollup could _start_ from the processed output of the serve lifecycle so we don't have to re-run userland PostCSS again.  In general, Rollup is doing a lot of inefficient work, would be great to refactor the whole thing from top to bottom. - see #453 for more context
    - not sure what exactly would was actionable here so not sure anything is needed here.  I think by disabling pre-rendering as part of #518 , this duplication concern would be avoided.
1. [x] [better lockstepping between serve and rollup for knowing all pages, e.g. sharing _graph.json_, but still try and keep it async?](https://github.com/ProjectEvergreen/greenwood/issues/327) 
    -  not sure what exactly would was actionable here, assuming the `input` for Rollup could just point to _src/_>. Nothing worth tracking for now, IMO.
1. [x] [Query Sorting / Filtering](https://github.com/ProjectEvergreen/greenwood/issues/288) - do this server side and inject current page for grapqhl side / slimmer fetching?  do this through the client instead (filter out the "dead pages" at the response level)?'
    - I think this was meant to be along the lines of a getServerSideProps type thing from NextJS.  I think this is better captured as a 2.0 discussion topic instead of `window.location`.
1. [x] [unit testing development](https://github.com/ProjectEvergreen/greenwood/issues/47)
1. [x] Determine Node version and use of `import`, also `(node:82640) ExperimentalWarning: The fs.promises API is experimental`  
 _node_modules_[(https://twitter.com/pikapkg/status/1306643472804278272).
1. [x] https://www.reddit.com/r/node/comments/fidzj9/fdir_the_fastest_directory_crawler_for_nodejs_10k - not sure if the extra dep is worth it, tbh
1. [x] replace JSDOM with something lighter / faster? - does it matter, only used for tests? - refactored a bit in #497
1. [x] Can be run via `npx` (0CJS - zero config JS) - confirmed as part of #499
1. [x] Good First Issues for Hacktoberfest - N / A
    - `serve` task from the CLI
    - Add timestamp to guides
    - [Html prettifier / formatter post optimize](https://github.com/ProjectEvergreen/greenwood/issues/318)
    - _Since Acorn 8.0.0, `options.ecmaVersion` is required._ warning message
1. [x] GitHub backlog
    - [sourcemaps](https://github.com/ProjectEvergreen/greenwood/issues/319)
    - [Progressive Enhancement aka RedactJS](https://github.com/ProjectEvergreen/greenwood/issues/354) - more like progressive refinement?
    - [404 Plugin](https://github.com/ProjectEvergreen/greenwood/issues/240)
    - [Capture intended API design](https://github.com/ProjectEvergreen/greenwood/issues/209) - think post processors, HTML snippets, build lifecycles / orchestration
    - [External Sources](https://github.com/ProjectEvergreen/greenwood/issues/21)
    - [website refresh](https://github.com/ProjectEvergreen/greenwood/issues/325)
    - [home page / project marketing](https://github.com/ProjectEvergreen/greenwood/issues/268)
1. [x] Validate if better HTML rendering is possible for second case - https://github.com/thegreenhouseio/www.thegreenhouse.io/issues/133
    - No special indentation
    - No restrictions on whitespace
    - Settings?
1. [x] 1.0 Release Blog Post on project evergreen website
    - [GitHub pages](https://github.com/ProjectEvergreen/greenwood/issues/320)
    - do anything about CEA? - will add SPA support to Greenwood
    - https://www.youtube.com/watch?v=7yA7BGos2KQ
    - ~~[netlify analytics](https://github.com/ProjectEvergreen/greenwood/issues/405)~~
1. [x] Make sure all TODOs in code are resolved or tracked - done in #497
1. [x] [**eslint-plugin-lint**](https://github.com/43081j/eslint-plugin-lit) - N / A
1. [x] refactor develop vs prod server and to use standard plugins (if applicable) - https://github.com/ProjectEvergreen/greenwood/issues/519
1. [x] refactor liveServer into its own plugin.  also, how to handle reloading for custom files e.g. _.foo_? - done as part of #471
1. [x] we need to add support for more font / icon types - https://github.com/ProjectEvergreen/greenwood/issues/521
1. [x] should Greenwood standard plugins be pushed to plugins array? - https://github.com/ProjectEvergreen/greenwood/issues/520
1. [x] Handle [inline `<script>` and `<style>` tags for `develop`](https://github.com/ProjectEvergreen/greenwood/pull/472#issuecomment-784666546) (and `prod`, of course).  Not sure if this can wait until post 1.0, but putting it on this roadmap for now. - https://github.com/ProjectEvergreen/greenwood/issues/534
1. [x] [Favicons plugin](https://github.com/ProjectEvergreen/greenwood/pull/474#issuecomment-784668512), and add one to getting started repo - https://github.com/ProjectEvergreen/greenwood/issues/194
1. [x] add support for [`@apollo/client` in _client.js_](https://github.com/ProjectEvergreen/greenwood/pull/487) - https://github.com/ProjectEvergreen/greenwood/issues/535
1. [x] need to ensure a clear distinction between when file extensions are / aren't needed when using `import` as the spec says they should always be there but for now, will make it so _node_modules_ can be bare, and can make an issue to add support for bare module specifier as a plugin - https://github.com/ProjectEvergreen/greenwood/issues/536
1. [x] should implement shared _package.json_ dependency walker for `importMaps` like in [**plugin-babel** / `regenerator-runtime`](https://github.com/ProjectEvergreen/greenwood/pull/481/files#diff-091732f9dd50157f76708e4269d733499ab1d22c84e56f3366a605c196366b9fR48) - https://github.com/ProjectEvergreen/greenwood/issues/537
1. [x] Currently need to have distinction between common and server client's until we can get to node v14 where we can use `import` on NodeJS code - https://github.com/ProjectEvergreen/greenwood/issues/532
1. [x] Make **plugin-graphql** [DRY for `importMap` and `rollupAlias`](https://github.com/ProjectEvergreen/greenwood/pull/487/files#diff-95f88393cf690e8e618da6f04158a085977afd1111119116a5166f08a28173c8R71) - https://github.com/ProjectEvergreen/greenwood/issues/538
1. [x] Should really implement [`server.stop` handling on `process` end](https://stackoverflow.com/questions/10522532/stop-node-js-program-from-command-line) - https://github.com/ProjectEvergreen/greenwood/issues/539
1. [x] try and avoid need for `createCache` round-robin? 🐦  - https://github.com/ProjectEvergreen/greenwood/issues/541
1. [x] Upgrade Node v14 - https://github.com/ProjectEvergreen/greenwood/issues/532
1. [x] ~~`eject` functionality~~ - [done](https://github.com/ProjectEvergreen/greenwood/pull/481/commits/73bf504715c8ea95932f663741fd64d12f59d8f3)!
1. [x] refactor **plugin-babel** [to not have to require manual installation of Babel deps](https://github.com/ProjectEvergreen/greenwood/pull/481/files#diff-091732f9dd50157f76708e4269d733499ab1d22c84e56f3366a605c196366b9fR48) like [**regenerator-runtime**](https://github.com/ProjectEvergreen/greenwood/pull/481/commits/4e1b9882df99ef7e49c1c0fd0e3f1f23b7056d55)
1. [x] top level pages don't merge their own `<title>` / `<meta>` tags (bug) - https://github.com/ProjectEvergreen/greenwood/issues/530
1. [x] improve coverage coverage - get these branches up!  and develop / serve tasks (+ server plugin / interface) - https://github.com/ProjectEvergreen/greenwood/issues/47
1. [x] imports from markdown frontmatter - https://github.com/ProjectEvergreen/greenwood/issues/529
1. [x] default command should be `build`? (for npx) - discussion - https://github.com/ProjectEvergreen/greenwood/discussions/528
1. [x] extract front matter contents from remark-frontmatter instead of separate frontmatter lib - https://github.com/ProjectEvergreen/greenwood/issues/531
1. [x] [rollup refactoring](https://github.com/ProjectEvergreen/greenwood/pull/497/commits/13838b44f5c81e00e229ce3c28d09fea85e6c28b) - hardcoded paths to _.greenwood/_, _public/_, path references, lot of (nested) loops - https://github.com/ProjectEvergreen/greenwood/issues/542
1. [x] copy assets from code, or general copy plugin - https://github.com/ProjectEvergreen/greenwood/issues/543
1. [x] In the Getting Started repo (npm), seeing this warning from Rollup - https://github.com/ProjectEvergreen/greenwood/issues/526
    ```sh
     @rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
    ```
1. [x] Should create an issue to track determining best optimization settings based on `mode` setting (at least by default).  Not all scripts are created equally, `preload` vs `prefetch`.  Router and GraphQL should probably use `prefetch`.  - https://github.com/ProjectEvergreen/greenwood/discussions/545
1. [x] Should make an issue to tracking adding per `<link>` / `<script>` overrides for `inline` and `static` optimization settings 
    - and then use that to optimize the site further (all inline vs selective uses of inline + static). - https://github.com/ProjectEvergreen/greenwood/issues/546
1. [x] Should make an issue to track adding per `<link>` / `<script>` overrides for preload or prefetch hints - https://github.com/ProjectEvergreen/greenwood/issues/546
1. [x] For partial route filename hashing, might require a refactoring of how HTML optimization happens, since when doing it _during_ serialization, you only assume that that one individual page has been built (since we are in a loop), and not the entire set, which means you can't use the graph to operate on the whole group to generate the same `<routes` for all pages.  Saved a [WIP branch here](https://github.com/ProjectEvergreen/greenwood/compare/enhancement/issue-354-restore-optimization-for-no-bundle...enhancement/issue-354-restore-optimization-for-no-bundle-route-partials-hashing-wip) and will track as a follow up discussion.  https://github.com/ProjectEvergreen/greenwood/issues/547
1. [x] Lighthouse test all pages - https://github.com/ProjectEvergreen/greenwood/issues/524
1. [x] Start a discussion about how to scale and apply a consistent design system over time to the website on pace for 1.0 release, to account for new features intended to be added that can be incorporated - https://github.com/ProjectEvergreen/greenwood/discussions/503
1. [x] Would be good start a discussion on investigating opportunities to make pre-rendering less dependent on Puppeteer like creating a [`HydrateElement`](https://github.com/thescientist13/hydrate-element) or looking into things like [Declarative Shadow DOM](https://web.dev/declarative-shadow-dom/) - https://github.com/ProjectEvergreen/greenwood/discussions/548
1. [x] Browserslist: caniuse-lite is outdated.  - https://github.com/ProjectEvergreen/greenwood/issues/525
    ```bash
    Please run: npx browserslist@latest --update-db
    ```
1. [x]  On content change, need to refresh graphql server in development and generate new _graph.json) (Use mutations to patch on the fly???) - https://github.com/ProjectEvergreen/greenwood/issues/549
1. [x] Make id unique to the route - https://github.com/ProjectEvergreen/greenwood/issues/552
1. [x] There is a double `item.item` in children query? - https://github.com/ProjectEvergreen/greenwood/issues/552
1. [x] Should only used Named exports  - https://github.com/ProjectEvergreen/greenwood/issues/551
1. [x] eliminate rollup plugin and use intercept? - https://github.com/ProjectEvergreen/greenwood/discussions/550
1. [x] Starter kits (discussion) - https://github.com/ProjectEvergreen/greenwood/discussions/523

## Post 1.0
These are tasks that can be worked on in the normal course of maintaining Greenwood and shouldn't require introducing any breaking changes.  Really valuable features should have the Post 1.0 Milestone assigned so that contributors can get some visibility into where help could be applied.  Post 1.0, we would probably moved to quarterly time boxed projects to track our status.

Maybe when we get to this point, we can all pick a couple tasks and start from there?  We can keep this list flexible.
1, Image loading plugin
1. [Explore strict mode for Greenwood website](https://github.com/ProjectEvergreen/greenwood/issues/410)
1. generate import map as JSON file and import from a <script> tag / path, and only once - #import
1. Pre bundle dependencies for development (a la snowpack) - #628 
1. Starter / Boilerplate / Generator (via `npx`), create a badge.io? - #523 
1. From the backlog
    - [open browser automatically](https://github.com/ProjectEvergreen/greenwood/issues/62)
    - [Performance automation](https://github.com/ProjectEvergreen/greenwood/issues/205)
    - [Differential Loading (IE11)](https://github.com/ProjectEvergreen/greenwood/issues/224)
      - https://github.com/ProjectEvergreen/greenwood/issues/198
      - https://github.com/ProjectEvergreen/greenwood/issues/196
    - [404 Plugin](https://github.com/ProjectEvergreen/greenwood/issues/240)
    - [Pre-processor plugins like SASS / SCSS](https://github.com/ProjectEvergreen/greenwood/issues/185). [TypeScript](https://github.com/ProjectEvergreen/greenwood/issues/46) support OOTB?  esbuild?  YML?
    - [Image processing](https://github.com/ProjectEvergreen/greenwood/issues/235) / lazy loading / general best practices (use of [`new URL` and `import.meta.url`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import.meta)?)
    - [CSS theming](https://github.com/ProjectEvergreen/greenwood/issues/111), constructable style sheets, inherited properties pierces the Shadow DOM
    - [PWA](https://github.com/ProjectEvergreen/greenwood/issues/194)
    - [console / debug logging](https://github.com/ProjectEvergreen/greenwood/issues/15)
    - [HMR](https://github.com/ProjectEvergreen/greenwood/issues/48)
    - [video series](https://github.com/ProjectEvergreen/greenwood/issues/380)
    - [html minification / validation](https://github.com/ProjectEvergreen/greenwood/issues/357)
    - [Update meta on SPA route change](https://github.com/ProjectEvergreen/greenwood/issues/306)
    - [Have develop mode pick random port](https://github.com/ProjectEvergreen/greenwood/issues/71)
    - [Build Lifecycles](https://github.com/ProjectEvergreen/greenwood/issues/184)
    - [one off page templates as JS](https://github.com/ProjectEvergreen/greenwood/issues/170)
    - [linting](https://github.com/ProjectEvergreen/greenwood/issues/106)
    - [Reduce duplicate data fetches](https://github.com/ProjectEvergreen/greenwood/issues/347)
1. [better local HTTP caching w/ `etag`](https://www.youtube.com/watch?v=3XkU_DXcgl0)
1. Single File Components (SFC)
1. Standalone SPA / Dynamic Apps (https://youtu.be/RVNC61rkFxw) - #589 
1. Rollup code splitting by template
1. preloading / async / defer modules (optimizations) - #545 
1. [Inline JS / CSS](https://developers.google.com/web/tools/puppeteer/articles/ssr)
1. [create a github action for puppeteer support](https://github.com/ProjectEvergreen/greenwood/pull/335#issuecomment-618044372)
1. proxy dev server for API calls - #577 
1. HTTP/2 for dev server - #632 
1. Restart graphql on file change? (Query, shelf) - #633 
1. Restart Koa server on file change (for local development) - #633 
1. Intelligent asset copy based on static analysis (e.g. favicon.ico)?  Or just default to copying all of _assets/_ ?
1. use _index.html_ as a default page template? - done!
1. Go [all in on unified](https://github.com/ProjectEvergreen/greenwood/issues/289)
1. `~` as a way to resolve node modules (as a plugin)?
1. dev server overlay (with build errors like what would be seen in the terminal) - #631 
1. Canary pipelines by decoupling repo from website and having GitHub action build the website with latest version of Greenwood as part of each build (own org at this point?) - #561 

## Thoughts for 2.0
Not necessarily breaking changes but also for more future facing ideas / features that might take a while or are just good for tracking at least?  I think this should all be tracked in Trello for now until we have hit the 1.0 release at least.  Then we can groom all of Trello and get it into a more refined GitHub backlog, which we should have by this time.  Might also be worth creating an RFC specific repo for these (unless GitHub discussions becomes available by then).

- [Web vital](https://github.com/stefanjudis/web-vitals-element) / [lighthouse reports in Puppeteer](https://pptr.dev/#?product=Puppeteer&version=v3.3.0&show=api-class-coverage) as a debugging tool / plugin?
- SSR component (a la Nuxt / Next) - #576 
- Caching in between builds / Incremental builds - maybe using GitHub actions + a greenwood api to build a single route / page on demand?
- [CSS Modules](https://github.com/css-modules/postcss-modules)
- evaluate [xm](https://github.com/giuseppeg/xm) / [posthtml](https://posthtml.org/) / [atomico](https://atomicojs.github.io/)
- web packaging?
- Streaming GraphQL data in / data out so server can start even if all data isn't ready yet as it becomes available / on demand
- Streaming / parallelized serialize and bundle phase, so we can rollup pages as they get serialized by **puppeteer**
- Would worker threads help our build process at all?
- Storybook support?
- serialize during development
- introduce a `getServerSideProps` type thing like from NextJS to help identify a page instead of `window.location === route`? - #576 / #626 
- decouple Rollup as the "initiator" for bundling so that Greenwood can just call out to any bundler instead of depending on Rollup to crawl all the HTML as the entry point.

## Tech Radar
Just interesting things to watch, maybe in Trello or dedicated GitHub repo.
- [Declarative SSR in Chrome, tweet from **Jason Miller**](https://twitter.com/_developit/status/1311012877428457472?)

**Comment by thescientist13** [comment]
I think we've got our mileage out of this ticket so going to close.  A lot of good information has now been captured in our issues and discussions.  v1.0.0 here we come!
